pkg/destroy/gcp/cloudcontroller.go

```diff
@@ -0,0 +1,201 @@
+package gcp
+
+import (
+	"fmt"
+
+	compute "google.golang.org/api/compute/v1"
+	"k8s.io/apimachinery/pkg/util/sets"
+)
+
+// listCloudControllerInstanceGroups returns instance groups created by the cloud controller.
+// It list all instance groups matching the cloud controller name convention.
+// https://github.com/openshift/kubernetes/blob/1e5983903742f64bca36a464582178c940353e9a/pkg/cloudprovider/providers/gce/gce_loadbalancer_naming.go#L33-L40
+// https://github.com/openshift/kubernetes/blob/1e5983903742f64bca36a464582178c940353e9a/pkg/cloudprovider/providers/gce/gce_clusterid.go#L210-L238
+func (o *ClusterUninstaller) listCloudControllerInstanceGroups() ([]cloudResource, error) {
+	filter := fmt.Sprintf("name eq \"k8s-ig--%s\"", o.cloudControllerUID)
+	return o.listInstanceGroupsWithFilter("items/*/instanceGroups(name,zone),nextPageToken", filter, nil)
+}
+
+// listBackendServicesForInstanceGroups returns backend services created by the cloud controller.
+// It list all backend services matching the cloud controller name convention that contain
+// only cluster instance groups.
+func (o *ClusterUninstaller) listBackendServicesForInstanceGroups(instanceGroups []cloudResource, installerInstanceGroups []cloudResource) ([]cloudResource, error) {
```

Comment:
why not a single list? compared to instanceGroups & installerInstanceGroups ?

Reply:
We are only destroying the cloud controller resources here... the installer created things are named by convention and will be destroyed in the normal workflow.

Reply:
hmm, not sure how that answers my single list of instancegroups comment??

Reply:
When iterating over the instance groups, it's checking which backend services belong. If a backend service that is not "on the list" is found, then we assume it's not a cluster resource. In some circumstances, these backend services contain the control plane instance groups, so we want to take them into consideration when determine what is and is not a cluster resource. Having two lists allows us to find the cloud controller resources with this consideration. By combining into a single list, this function would instead find any backend service attached to the cluster, not just the ones created by the Cloud Controller. This might be ok, but it's not the intention of the function.

Reply:
> https://github.com/openshift/installer/pull/2589/files/97cd4afee65d7d7f90a7b37502f7fb8e937be9b0#diff-fe15e1513995e1b091fe47a3c8d79d0eR25-R31

this function does not care if the instance groups are k8s-managed or installer-created, why request 2 different types of these as input.

Reply:
Disregard... `filter := "name eq \"a[0-9a-f]{30,50}\""` will exclude ClusterID named resources. This can be a single list and still function the same.

Reply:
Updated to be a single list.